### PR TITLE
Kinesis channel performance test report

### PR DESCRIPTION
**Pull request by tzununbekov**
I had to re-run tests since I expected to see the difference between GKE->AWS and EKS->AWS test environments.
[Rendered doc](https://github.com/triggermesh/test-infra/tree/kinesis-channel-report/perf/reports/20210215-kinesis-channel)

Follow up issue:
https://github.com/triggermesh/aws-kinesis-channel/issues/59

### CODE REVIEW

File: perf/reports/20210215-kinesis-channel/README.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+Bytes In      [total, mean]              0, 0.00
+Bytes Out     [total, mean]              364222464, 1785.84
+Success       [ratio]                    84.33%
+Status Codes  [code:count]               0:26107  202:171983  500:5860
+Error Set:
+500 Internal Server Error
+Post "http://kinesis-ch-dispatcher.perf-thrpt-receiver.svc.cluster.local": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
+```
+
+With the steam rate of 600 events per second channel could successfully handle 84.33% of events for 6 minutes before dispatcher timeouts exceeded the receiver's wait period.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:77]:
```suggestion
With the stream rate of 600 events per second channel could successfully handle 84.33% of events for 6 minutes before dispatcher timeouts exceeded the receiver's wait period. 
```

File: perf/reports/20210215-kinesis-channel/README.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+Requests      [total, rate, throughput]  300000, 500.00, 499.76
+Duration      [total, attack, wait]      10m0.286979935s, 9m59.997784283s, 289.195652ms
+Latencies     [mean, 50, 95, 99, max]    11.104268ms, 9.131272ms, 21.475218ms, 40.570542ms, 993.733317ms
+Bytes In      [total, mean]              0, 0.00
+Bytes Out     [total, mean]              614400000, 2048.00
+Success       [ratio]                    100.00%
+Status Codes  [code:count]               202:300000
+```
+
+In the first two tests Kinesis channel successfully sustained throughput of 300 and 500 events per second for 10 minutes with average latency around 10ms.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:60]:
```suggestion
In the first two tests, the Kinesis channel successfully sustained throughput of 300 and 500 events per second for 10 minutes with average latency around 10ms.
```

File: perf/reports/20210215-kinesis-channel/README.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+Success       [ratio]                    66.01%
+Status Codes  [code:count]               0:30246  202:61228  500:1275
+Error Set:
+500 Internal Server Error
+Post "http://kinesis-ch-dispatcher.perf-thrpt-receiver.svc.cluster.local": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
+```
+
+700 events per second demonstrated the expected drop of dispatching performance - 66.01% of success for 2.5 minutes.
+
+Monitoring data showed that the most probable reason for observed behavior is suboptimal memory management in the Kinesis channel dispatcher. Resources other than RAM had linear graphs with no spikes throughout all testing.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:96]: These first three tests are representing data ingesting into the stream and data reads from the stream to get to the receiver, correct ?

File: perf/reports/20210215-kinesis-channel/README.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+* A single instance of sender and receiver are able to sustain the load of 12,000+ events per second which is far above expected channel performance and Kinesis shard [limits](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html). Thus, one sender and receiver were enough to create a sufficient data test stream.
+
+---
+
+## Results
+
+### Sustained load
+
+For sustained loading tests, we used [cegen](../../tools/cegen/) to generate CloudEvents with a fixed size of 2KiB, piped to `vegeta` attacker configured to send these events to a receiver. We ran multiple tests with different events stream rates to find the exact throughput that can be sustained by the channel. AWS Kinesis Channel had 3 shards in all test cases.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:33]: I would like to see a basic test of ingesting data into the stream using vegeta in order to verify that we can ingest at around 3,000 events/sec into the stream.

File: perf/reports/20210215-kinesis-channel/README.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+700 events per second demonstrated the expected drop of dispatching performance - 66.01% of success for 2.5 minutes.
+
+Monitoring data showed that the most probable reason for observed behavior is suboptimal memory management in the Kinesis channel dispatcher. Resources other than RAM had linear graphs with no spikes throughout all testing.
+
+![Memory consumption](.assets/pods-memory-usage-sustained.png)
+_300, 500 and 700 events/sec respectively_
+
+
+AWS Kinesis channel that was used during the tests had 3 shards which in total should be capable of up to 3 MB/sec (or 3000 records/sec) of "put" operations and up to 30 MB/sec (or 30000 records/sec) for "get" operations. AWS monitoring showed that we were far from reaching these limits:
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:102]: to remove the "should" in this first sentence we need to run an ingesting test alone.

File: perf/reports/20210215-kinesis-channel/README.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+_250,221,487/1,887,436,800 bytes read_ 
+
+![AWS Kinesis Write](.assets/aws-kinesis-put-sustained.png)
+_250,351,669/943,718,400 byte write_
+
+
+### Burst load
+
+To get the complete picture of the channel performance we ran an additional set of tests with a series of burst loads. Test script runs a succession of attacks at different rates in periods of 5s, ranging from 1 event/sec to 79,432 events/sec (theoretical),
+with a payload of 2 KiB. AWS Kinesis stream backed the channel was configured with different numbers of shards for the same test sets. In general, the channel performance was not affected by the number of the shards as the limiting factor is the dispatching process.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:114]:
```suggestion
with a payload of 2 KiB. The AWS Kinesis stream backing the channel was configured with different numbers of shards for the same test sets. In general, the channel performance was not affected by the number of the shards as the limiting factor is the dispatching process.
```

File: perf/reports/20210215-kinesis-channel/README.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+Although these graphs are not that easy to read, from the receiver reports we see the numbers that we could expect after sustained load tests - maximum events receive rate is in between 500 and 750 events per second. Vegeta report, in its turn, shows that the send operations under 500 event/sec rate have latencies in between 10 and 100ms, and then it drastically starts to spike out above 1 second.
+
+![Pods CPU usage](.assets/pods-cpu-usage.png)
+![Pods memory usage](.assets/pods-memory-usage.png)
+![Dispatcher Network usage](.assets/dispatcher-network.png)
+_Typical resources usage profile during tests_
+
+Kinesis channel asynchronous nature makes it hard to point out the exact component of the dispatcher that creates current performance limitation without additional debugging. The fact that the "get" operation reads batches of records while "put" writes individual events may create performance inconsistence between sending part of the dispatcher and its receiveng counterpart.
+
+In a conclusion, we recommend limiting event flow through the Kinesis channel to 500 events/sec until further improvements are made to the dispatcher controller.
```
Comment (sebgoa) [inline perf/reports/20210215-kinesis-channel/README.md:143]:
```suggestion
In a conclusion, we recommend limiting the event flow through the Kinesis channel to 500 events/sec until further improvements are made to the dispatcher controller.
```

**Comment by sebgoa** [comment]
Thanks for this Timur.

I do like the final recommendation but we should definitely investigate further to understand why we are far from the 3,000 events/sec that 3 shards could ingest.

Can you follow up with an issue in the kinesis channel repo ?

thanks

**Review by sebgoa — CHANGES_REQUESTED** [review]

File: perf/reports/20210215-kinesis-channel/README.md (hunk 3 above)
Comment (tzununbekov) [reply perf/reports/20210215-kinesis-channel/README.md:96]: These numbers are about ingesting events into the Kinesis stream. The receiver end report graph shows the number of received events per second and it is very dispersed: 
chart

I'll include these graphs in the update.

File: perf/reports/20210215-kinesis-channel/README.md (hunk 5 above)
Comment (tzununbekov) [reply perf/reports/20210215-kinesis-channel/README.md:102]: will do

**Comment by tzununbekov** [comment]
@sebgoa I added a couple of fixes to the controller, re-run tests, and updated the report. Please take another look

**Comment by sebgoa** [comment]
thanks for the improvements, happy to see we can reach the shard limits.

**Review by sebgoa — APPROVED** [review]

**Pull request merged**